### PR TITLE
Add license agreement checking for yast2_migration

### PR DESCRIPTION
**Pull request by chcao**
Add license agreement checking for yast2 migration cases. Based on bsc#1185808, license agreement is not printed when doing yast2 migration, now the bug has been fixed, we should add the license agreement checking and accept the license to the automation testing.

- Related ticket: https://progress.opensuse.org/issues/92620
- Verification run: 
https://openqa.suse.de/tests/6464530#step/yast2_migration/8 (sles s390x textmode)
https://openqa.suse.de/tests/6464552#step/yast2_migration/19 (sles x86_64 gnome)
https://openqa.suse.de/tests/6464536#step/yast2_migration/19 (sled x86_64 gnome)
https://openqa.suse.de/tests/6464537#step/yast2_migration/7 (hpc x86_64 textmode)
https://openqa.suse.de/tests/6464532#step/yast2_migration/20 (leap to sle x86_64 gnome)

### CODE REVIEW

File: tests/migration/online_migration/yast2_migration.pm [hunk 1]
```diff
@@ -220,10 +220,13 @@ sub run {
             send_key 'alt-t';
         }
     }
-    assert_screen ['yast2-migration-installupdate', 'yast2-migration-proposal'], 700;
-    if (match_has_tag 'yast2-migration-installupdate') {
+    if (check_screen('yast2-migration-installupdate', 700)) {
```
Comment (coolgw) [inline tests/migration/online_migration/yast2_migration.pm:223]: when yast2-migration-proposal can be match, in this logic seems SKIP the match action.

Comment (rfan1) [inline tests/migration/online_migration/yast2_migration.pm:223]: Why don't we use assert_screen here?

Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm:223]: Because for some cases, there's no yast2-migration-installupte.

Comment (rfan1) [reply tests/migration/online_migration/yast2_migration.pm:223]: Got it, then no problem. can you please add some comments there? then as a new learner, I am a little confusing why you check this screen.

Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm:223]: The license agreement is shown between yast2-migration-installupdate and yast2-migration-proposal.  While from the test results, the yast2-migration-proposal and the yast2-migration-summary are in the same page and match different needle, and the cases match the needle of yast2-migration-summary. So I remove the yast2-migration-propoasl here.

Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm:223]: ok, no problem and thanks.

**Review by tinawang123 — APPROVED** [review]
LGTM

**Review by rfan1 — APPROVED** [review]
LGTM

File: tests/migration/online_migration/yast2_migration.pm (hunk 1 above)
Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm]: Done with adding a comment here, thanks.

File: tests/migration/online_migration/yast2_migration.pm [hunk 2]
```diff
@@ -220,10 +220,13 @@ sub run {
             send_key 'alt-t';
         }
     }
-    assert_screen ['yast2-migration-installupdate', 'yast2-migration-proposal'], 700;
-    if (match_has_tag 'yast2-migration-installupdate') {
+    if (check_screen('yast2-migration-installupdate', 700)) {    # Not all cases have install update message
```
Comment (lemon-suse) [inline tests/migration/online_migration/yast2_migration.pm:223]: Why not continue using the multi-tag?

**Comment by lemon-suse** [comment]
Please add comments for each verification job in PR description and add more description for this PR such as history, why need it and how you fix it.

File: tests/migration/online_migration/yast2_migration.pm (hunk 2 above)
Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm:223]: The current situation doesn't fit to use multi-tag any more: the license agreement is shown between yast2-migration-installupdate and yast2-migration-proposal/summary, and the license must be shown, otherwise we should file bug. The flow here is:
installupdate (maybe have it);
license agreement (must have it);
migration proposal (must have it, but it is in the same page with migration_summary, two tags in the same page)

Comment (okurz) [inline tests/migration/online_migration/yast2_migration.pm:223]: Please find a better solution than wasting 700s (!). I don't see why your situation would deny the use of multi-tag.

**Review by okurz — CHANGES_REQUESTED** [review]

File: tests/migration/online_migration/yast2_migration.pm (hunk 2 above)
Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm:223]: I'm testing more cases to reduce the time here. 
For the multi-tag, the license agreement I added here is appeared between yast2-migration-installupdate and yast2-migration-proposal/summary, and the license must be shown, it is not optional. The flow here is:
firstly, installupdate (maybe have it, some cases don't have it);
then, license agreement (all the cases must print it, otherwise we should file a bug);
and then migration proposal (must have it, but it is in the same page with migration_summary, two tags in the same page)

Comment (okurz) [reply tests/migration/online_migration/yast2_migration.pm:223]: Sorry, I don't see why you should not be able to use a better approach. This brought me to the idea to introduce an automatic code style check: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12821

**Review by okurz — CHANGES_REQUESTED** [review]

File: tests/migration/online_migration/yast2_migration.pm (hunk 2 above)
Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm:223]: If using multi-tag, when it matches one of the three tags, the case will pass, right? In this situation, there are three tags, it should match tag1, tag2 and tag3; or it should match tag2 and tag3, so multi-tag is not suitable here.

Comment (okurz) [reply tests/migration/online_migration/yast2_migration.pm:223]: > so multi-tag is not suitable here.

that's not correct. You can always find a way to express final states that you want to reach. A sane human operator would also not "wait for 1 or 2 or until 12 minutes have passed"

Comment (okurz) [reply tests/migration/online_migration/yast2_migration.pm]: In https://openqa.nue.suse.com/tests/6267218#step/yast2_migration/31 I can see that the test fails expecting "yast2-migration-installupdate" or "yast2-migration-proposal" but instead a license dialog shows up which can be looked for with "yast2_migration-license-agreement". Of course after handling the license agreement you are not done with the migration so I assume after accepting the license you still want to look for "yast2-migration-installupdate" or "yast2-migration-proposal", right? So you could use something like

```perl
assert_screen [qw(yast2-migration-installupdate yast2-migration-proposal yast2_migration-license-agreement)];
if (match_has_tag 'yast2_migration-license-agreement') {
    send_key 'alt-a';
    assert_screen [qw(yast2-migration-installupdate yast2-migration-proposal)];
}
…
```

https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/x11/kontact.pm#L21 provides a good elaborate example of how one can manage different dialogs that optionally pop up and in not defined order.

Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm]: Thanks for your advise. I'm going to have a try and update here.

Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm]: The case doesn't fit to use qw(), let me explain it here: installupdate is shown BEFORE license, while not all cases have installupdate message, so we need to use "if" on installupdate; License is must be shown for all cases, so we shouldn't use "if" for it. BTW, in https://openqa.nue.suse.com/tests/6267218#step/yast2_migration/31, we can see the license is shown, though we accept it, the next page is not installupdate, it would be proposal(same page with summary); For proposal, it is the same page with summary.

Comment (okurz) [reply tests/migration/online_migration/yast2_migration.pm]: Sorry, I don't understand your explanation why the quoted word list is not applicable. What has `qw()` have to do with the test flow? Have you looked at the kontact example with the loop? I think that could help

**Comment by okurz** [comment]
please update the verification runs

File: tests/migration/online_migration/yast2_migration.pm (hunk 2 above)
Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm]: Yes, you are right, finally I found a solution, thanks for your help.

**Comment by chcao** [comment]
> please update the verification runs

I update the verification runs, please help to review, thanks.

File: tests/migration/online_migration/yast2_migration.pm [hunk 3]
```diff
@@ -220,10 +220,20 @@ sub run {
...
         send_key 'alt-y';
+        assert_screen 'yast2_migration-license-agreement', 60;
+        wait_screen_change { send_key "alt-a" };
+        assert_screen 'yast2_migration-license-agreenment-accepted', 60;
+        send_key "alt-n";
     }
+    if (match_has_tag 'yast2_migration-license-agreement', 150) {
+        wait_screen_change { send_key "alt-a" };
+        assert_screen 'yast2_migration-license-agreenment-accepted', 60;
+        send_key "alt-n";
```
Comment (okurz) [inline tests/migration/online_migration/yast2_migration.pm:234]: as you use this same code block twice I suggest to extract that into a function with an explanatory name and call in both places

**Review by okurz — CHANGES_REQUESTED** [review]

**Review by okurz — APPROVED** [review]

File: tests/migration/online_migration/yast2_migration.pm (hunk 3 above)
Comment (chcao) [reply tests/migration/online_migration/yast2_migration.pm]: I update the code, and the verification cases are running, I'll update the results later, thanks for your review.

**Comment by okurz** [comment]
Thank you. The current version looks much cleaner and I am happy :)

**Comment by chcao** [comment]
> Thank you. The current version looks much cleaner and I am happy :)

I’m happy too :)

**Review by lemon-suse — APPROVED** [review]
